src/AsyncAPI.Saunter.Generator.Cli/ToFile/AsyncApiDocumentExtractor.cs

```diff
@@ -0,0 +1,72 @@
+﻿using LEGO.AsyncAPI.Models;
+using LEGO.AsyncAPI.Readers;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Saunter;
+using Saunter.Serialization;
+
+namespace AsyncAPI.Saunter.Generator.Cli.ToFile;
+
+internal interface IAsyncApiDocumentExtractor
+{
+    IEnumerable<(string name, AsyncApiDocument document)> GetAsyncApiDocument(IServiceProvider serviceProvider, string[] requestedDocuments);
+}
+
+internal class AsyncApiDocumentExtractor(ILogger<AsyncApiDocumentExtractor> logger) : IAsyncApiDocumentExtractor
+{
+    private IEnumerable<string> GetDocumentNames(string[] requestedDocuments, AsyncApiOptions asyncApiOptions)
+    {
+        var documentNames = requestedDocuments ?? asyncApiOptions.NamedApis.Keys;
+        if (documentNames.Count == 0)
+        {
+            if (asyncApiOptions.AssemblyMarkerTypes.Any())
+            {
+                documentNames = [null]; // null marks the default, unnamed, document
+            }
+            else
+            {
+                logger.LogCritical($"AsyncAPI documents found. Known named document(s): {string.Join(", ", asyncApiOptions.NamedApis.Keys)}.");
+            }
+        }
+        return documentNames;
+    }
+
+    public IEnumerable<(string name, AsyncApiDocument document)> GetAsyncApiDocument(IServiceProvider serviceProvider, string[] requestedDocuments)
+    {
+        var documentProvider = serviceProvider.GetService<IAsyncApiDocumentProvider>();
+        var asyncApiOptions = serviceProvider.GetService<IOptions<AsyncApiOptions>>().Value;
+        var documentSerializer = serviceProvider.GetRequiredService<IAsyncApiDocumentSerializer>();
+
+        foreach (var documentName in GetDocumentNames(requestedDocuments, asyncApiOptions))
+        {
+            if (documentName == null || !asyncApiOptions.NamedApis.TryGetValue(documentName, out var prototype))
+            {
+                prototype = asyncApiOptions.AsyncApi;
+            }
+
+            var schema = documentProvider.GetDocument(asyncApiOptions, prototype);
+            var asyncApiSchemaJson = documentSerializer.Serialize(schema);
+            var asyncApiDocument = new AsyncApiStringReader().Read(asyncApiSchemaJson, out var diagnostic);
```

Comment:
note that this technically read 2.6 and not 2.4 which has 'some' discrepancies.
Not that there is much to do about it.
Hopefully ill be able to get some time in to work on moving saunter forward to V3 (which is now supported by AsyncAPI.Net)